gcc/d/d-decls.cc

```diff
@@ -101,7 +101,38 @@ VarDeclaration::toSymbol()
 	  return csym;
 	}
 
-      csym = new Symbol();
+      // Use only same Symbol for VarDeclarations with same mangle
+      if (!this->mangleOverride && isDataseg () && !(storage_class & STCextern))
+	{
+	  static hash_map<nofree_string_hash, Symbol *> duplicate_map(251);
+	  const char *vname = mangle(this);
+	  bool existed;
+
+	  Symbol **entry = &duplicate_map.get_or_insert(vname, &existed);
+	  if (!existed)
+	    {
+	      csym = new Symbol();
+	      *entry = csym;
+	    }
+	  else
+	    {
+	      csym = *entry;
+	      if (!csym->Sloc.equals(loc))
+		{
+		  error ("Found different variables with identical mangle: %s"
+		    " at %s and %s", vname, loc.toChars(), csym->Sloc.toChars());
```

Comment:
> Maybe I misunderstood, but I don't think this would work. As duplicated template instances are also created in the same binding-level this would also error for template duplicates.

Well, we can tell the difference between normal functions and templates in both D AST and GCC's.  Whether it is `fd->isInstantiated()` or `DECL_COMDAT(fd)`